### ISSUE TITLE
More informative `Exception`/Error messages in converter

### ISSUE THREAD

**Issue by heinezen** [SFTtech/openage#1344]
**Required skills:** Python

**Difficulty:** Medium

In AoE2's `.dat` format most associations and assignments of properties are done by IDs (e.g. unit has ability with ID X). The openage converter uses these IDs to lookup the associated openage API property and then map the values from AoE2's `.dat` structure  to the corresponding API object's member values. In short, every property from AoE2 needs to be manually mapped to an openage API property. As such, most of the runtime errors are actually lookup errors (usually Python's `KeyError`) that occur when an AoE2 property was not mapped to an openage API property.

The goal of this task is to make these errors more informative by catching the generic `KeyError` from Python and re-raising it with a better message. For example, we can improve the error message by specifying:
  * Context of the property (i.e. the property is: ability, resource, unit, tech, ...)
  * Data type that is looked up (function, object, string, other ID, ...)
    
**Example**

generic message: 

```
File "openage/convert/processor/conversion/de2/tech_subprocessor.py", line 276, in resource_modify_effect
    upgrade_func = DE2TechSubprocessor.upgrade_resource_funcs[resource_id]
KeyError: 208
```

better message:

```
File "openage/convert/processor/conversion/de2/tech_subprocessor.py", line 276, in resource_modify_effect
    upgrade_func = DE2TechSubprocessor.upgrade_resource_funcs[resource_id]
KeyError: No subprocessor function found for handling upgrade of civ resource: 208
```

Use try-except statement where these lookups could be thrown. Remember to use `raise ... from` to preserve the stack trace of the initial `KeyError`.

**Further reading**:

* [Converter code](https://github.com/SFTtech/openage/blob/master/openage/convert/)
* [Converter architecture](https://github.com/SFTtech/openage/blob/master/doc/code/converter/architecture_overview.md)
* [Converter workflow](https://github.com/SFTtech/openage/blob/master/doc/code/converter/workflow.md)

**Comment by duanqn** [comment]
When you say 

> most of the runtime errors are actually lookup errors

Are those errors code defects, or just because the user does not have the correct AoE2 assets?

**Comment by heinezen** [comment]
Most of them are code defects or missing implementations, e.g.
* For somehing like tech effects, the converter gets the effect ID and looks up a corresponding handler function in a dict. If the ID is not found, the `KeyError` should be caught and raise a `NotImplementedError`.
* KeyErrors when looking up unit ID or tech IDs are almost always the converters fault
* Names for converted nyan objects are provided by the converter and a key error here indicates that the name must be added by us

The correct AoE2 assets should be detected before the conversion starts.

**Comment by tsherman181** [comment]
Hi all! I'm a first time contributor and I am going to try to solve this issue. Very new to open source contributing but hoping I can help here!

**Comment by heinezen** [comment]
@tsherman181 Let us know if you have any problems!

**Comment by anupriyakkumari** [comment]
Hi, I would like to take this up, please.

**Comment by heinezen** [comment]
@anupriyakkumari Hey yes this is free for implementation :)

**Comment by jsutcodes** [comment]
@heinezen is this issue still up for implementation? Newer to open source, but would love to contribute to the project.

**Comment by TheJJ** [comment]
Yes it is! We would hopefully have closed the issue otherwise :)

**Comment by jsutcodes** [comment]
@TheJJ ok, great I'll work the issue if you'd like to assign it to me.

**Comment by heinezen** [comment]
@jsutcodes No assignment necessary, you can just work on it and open a PR :)

**Comment by yoyonel** [comment]
Hi, interresting project, i like it 😄

All Python codes i read seems to be integrated in your CMake build strategy (i'm not familiar with this concept, using a building/compiling tool for scripting/interpreting language).

Do you have any (unit) tests for your Python applications, tools, services ?
In a quick exploration and search in this repo, i don't see any file with `test` in the filename for Python ... i miss something ?

Thanks in advance for your response :-)

**Comment by yoyonel** [comment]
My bad you've seems to have something looking like tests for Python code, here for example : 

![Image](https://github.com/user-attachments/assets/11f62a73-54e3-4408-b967-4e083988dfdc)

It's a integration tests ? I don't see any calls to [pytest](https://docs.pytest.org/en/stable/) or [unittest](https://docs.python.org/3/library/unittest.html) or another frameworks for testing (Python applications/libs/tools/...)

**Comment by TheJJ** [comment]
we created our own testing "framework", which is in `openage/testing` and `libopenage/testing` :)
